proof/invariant-abstract/AARCH64/ArchFinalise_AI.thy

```diff
@@ -117,32 +182,35 @@ lemma set_asid_pool_unmap:
    set_asid_pool poolptr (pool(asid_low_bits_of asid := None))
    \<lbrace>\<lambda>rv s. vspace_for_asid asid s = None \<rbrace>"
   unfolding set_asid_pool_def
-  apply (wpsimp wp: set_object_wp)
-  sorry (* FIXME AARCH64
-  by (simp add: pool_for_asid_def vspace_for_asid_def vspace_for_pool_def obind_def in_omonad
-         split: option.splits) *)
+  apply (wp set_object_wp)
+  by (simp add: pool_for_asid_def entry_for_asid_def entry_for_pool_def vspace_for_asid_def
+                vspace_for_pool_def obind_def in_omonad
+         split: option.splits)
+
+crunches invalidate_asid_entry
+  for pool_for_asid[wp]: "\<lambda>s. P (pool_for_asid asid s)"
+  (simp: pool_for_asid_def)
 
 lemma delete_asid_unmapped:
   "\<lbrace>\<lambda>s. vspace_for_asid asid s = Some pt\<rbrace>
    delete_asid asid pt
    \<lbrace>\<lambda>_ s. vspace_for_asid asid s = None\<rbrace>"
   unfolding delete_asid_def
   apply (simp cong: option.case_cong)
-  sorry (* FIXME AARCH64
-  apply (wpsimp wp: set_asid_pool_unmap)
+  apply (wpsimp wp: set_asid_pool_unmap | wp (once) hoare_drop_imps)+
   apply (clarsimp simp: vspace_for_asid_def pool_for_asid_def vspace_for_pool_def
-                        obind_def in_omonad obj_at_def
+                        obind_def in_omonad entry_for_asid_def entry_for_pool_def
                  split: option.splits)
-  done *)
+  by (meson asid_pool_entry.exhaust_sel)
 
 lemma set_pt_tcb_at:
   "\<lbrace>\<lambda>s. P (ko_at (TCB tcb) t s)\<rbrace> set_pt a b \<lbrace>\<lambda>_ s. P (ko_at (TCB tcb) t s)\<rbrace>"
   by (wpsimp simp: set_pt_def obj_at_def wp: set_object_wp)
 
 lemma set_vcpu_tcb_at_arch: (* generalise? this holds except when the ko is a vcpu *)
   "\<lbrace>\<lambda>s. P (ko_at (TCB tcb) t s)\<rbrace>
-      set_vcpu p v
-        \<lbrace>\<lambda>_ s. P (ko_at (TCB tcb) t s)\<rbrace>"
+   set_vcpu p v
+   \<lbrace>\<lambda>_ s. P (ko_at (TCB tcb) t s)\<rbrace>"
```

Comment:
Very unimportant, but with this lemma already changing it could also be changed to invariant notation.